File: app/src/main/java/torille/fi/lurkforreddit/data/remote/RedditRemoteDataSource.java
package torille.fi.lurkforreddit.data.remote;

import android.support.annotation.NonNull;
import android.support.v4.util.Pair;
import android.text.TextUtils;

import com.google.gson.stream.JsonReader;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;
import torille.fi.lurkforreddit.data.RedditDataSource;
import torille.fi.lurkforreddit.data.RedditService;
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentListing;
import torille.fi.lurkforreddit.data.models.jsonResponses.MultiredditListing;
import torille.fi.lurkforreddit.data.models.jsonResponses.PostListing;
import torille.fi.lurkforreddit.data.models.jsonResponses.SubredditListing;
import torille.fi.lurkforreddit.data.models.view.Comment;
import torille.fi.lurkforreddit.data.models.view.Post;
import torille.fi.lurkforreddit.data.models.view.PostAndComments;
import torille.fi.lurkforreddit.data.models.view.SearchResult;
import torille.fi.lurkforreddit.data.models.view.Subreddit;
import torille.fi.lurkforreddit.di.scope.RedditScope;
import torille.fi.lurkforreddit.utils.CommentsStreamingParser;
import torille.fi.lurkforreddit.utils.Store;
import torille.fi.lurkforreddit.utils.TextHelper;

/**
 * Created by eva on 23.4.2017.
 */
@RedditScope
public class RedditRemoteDataSource implements RedditDataSource {

    private final RedditService.Reddit mRedditApi;
    private final Store mSettingsStore;
    private final CommentsStreamingParser mCommentsStreamingParser;

    @Inject
    RedditRemoteDataSource(@NonNull RedditService.Reddit api,
                           @NonNull Store store,
                           @NonNull CommentsStreamingParser commentsStreamingParser) {
        mRedditApi = api;
        mSettingsStore = store;
        mCommentsStreamingParser = commentsStreamingParser;
    }

    @Override
    public Observable<List<Subreddit>> getSubreddits() {
        Timber.d("Fetching subs!");
        Observable<SubredditListing> subreddits;
        if (mSettingsStore.isLoggedIn()) {
            Timber.d("Was logged in, getting personal subreddits");
            subreddits = mRedditApi.getMySubreddits(200);

            return Observable.zip(fetchSubreddits(subreddits), getUserMultireddits(),
                    (subreddits1, subreddits2) -> {
                        Collections.addAll(subreddits1, subreddits2.toArray(new Subreddit[0]));
                        return subreddits1;
                    })
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread());
        } else {
            Timber.d("Was not logged in, getting default subreddits");
            subreddits = mRedditApi.getDefaultSubreddits(100);
            return fetchSubreddits(subreddits);

        }
    }

    @Override
    public Observable<Pair<String, List<Post>>> getSubredditPosts(@NonNull String subredditUrl) {

        return mRedditApi
                .getSubreddit(subredditUrl)
                .observeOn(Schedulers.computation())
                .map(funcFormatPostData);
    }

    @Override
    public Observable<Pair<String, List<Post>>> getMoreSubredditPosts(@NonNull String subredditUrl, @NonNull String nextpageId) {
        return mRedditApi.getSubredditNextPage(subredditUrl, nextpageId)
                .observeOn(Schedulers.computation())
                .map(funcFormatPostData);
    }

    @Override
    public void refreshData() {
        //empty
    }

    @Override
    public Observable<PostAndComments> getCommentsForPost(@NonNull String permaLinkUrl,
                                                          final boolean isSingleCommentThread) {

        return mRedditApi.getComments(permaLinkUrl)
                .map(responseBody -> {
                    try (InputStream stream = responseBody.byteStream();
                         InputStreamReader in = new InputStreamReader(stream, "UTF-8");
                         JsonReader reader = new JsonReader(in)) {

                        return mCommentsStreamingParser
                                .readCommentListingArray(reader);
                    }
                })
                .observeOn(Schedulers.computation())
                .map(commentListings -> {
                    Observable<List<CommentListing>> commentListingsObservable = Observable.fromArray(commentListings);
                    return Observable.zip(getPost(commentListingsObservable),
                            getFormattedComments(commentListingsObservable,
                                    isSingleCommentThread),
                            PostAndComments::create).blockingSingle();
                });
    }

    @Override
    public Observable<List<Comment>> getMoreCommentsForPostAt(@NonNull List<String> childCommentIds,
                                                              @NonNull String linkId,
                                                              final int commentLevel) {
        return mRedditApi.getMoreComments(linkId,
                TextUtils.join(",", childCommentIds),
                "json")
                .observeOn(Schedulers.computation())
                .map(responseBody -> {
                    try (InputStream stream = responseBody.byteStream();
                         InputStreamReader in = new InputStreamReader(stream, "UTF-8");
                         JsonReader reader = new JsonReader(in)) {

                        return mCommentsStreamingParser
                                .readMoreComments(reader);
                    }
                })
                .map(commentChildren ->
                        TextHelper.flattenAdditionalComments(commentChildren, commentLevel));
    }

    @Override
    public Observable<Pair<String, List<SearchResult>>> getSearchResults(@NonNull String query) {

        return mRedditApi.searchSubreddits(query, "relevance")
                .observeOn(Schedulers.computation())
                .map(formatSearchData)
                .doOnError(throwable -> {
                    return ;
                });
    }

    @Override
    public Observable<Pair<String, List<SearchResult>>> getMoreSearchResults(@NonNull String query, @NonNull String afterId) {
        return mRedditApi.searchSubredditsNextPage(query, "relevance", afterId)
                .observeOn(Schedulers.computation())
                .map(formatSearchData);
    }

    private Observable<List<Subreddit>> getUserMultireddits() {
        return mRedditApi.getUserMultireddits()
                .observeOn(Schedulers.computation())
                .map(Observable::fromArray)
                .flatMap(multiredditListingObservable -> multiredditListingObservable.map(MultiredditListing::multireddit))
                .map(multireddit -> Subreddit.builder()
                        .setUrl(multireddit.pathUrl())
                        .setBannerUrl(null)
                        .setId(multireddit.name())
                        .setKeyColor(multireddit.keyColor())
                        .setDisplayName(multireddit.displayName())
                        .build())
                .toList().toObservable();
    }

    private Observable<List<Subreddit>> fetchSubreddits(Observable<SubredditListing> listingObservable) {
        return listingObservable
                .subscribeOn(Schedulers.computation())
                .map(subredditListing -> Observable.fromIterable(subredditListing.data().children()))
                .flatMap(TextHelper::formatSubreddit)
                .toSortedList((subreddit1, subreddit2) ->
                        subreddit1.displayName().compareToIgnoreCase(subreddit2.displayName())
                ).toObservable();
    }

    private Observable<Post> getPost(final Observable<List<CommentListing>> observable) {
        return observable
                .map(commentListings -> Observable.fromArray(commentListings.get(0)
                        .commentData()
                        .commentChildren()
                        .get(0)
                        .originalPost()))
                .map(TextHelper.funcFormatPostDetails).blockingSingle();
    }

    private Observable<List<Comment>> getFormattedComments(Observable<List<CommentListing>> observable, final boolean isSingleCommentThread) {

        return observable
                .subscribeOn(Schedulers.computation())
                .map(commentListings ->
                        commentListings.get(1)
                                .commentData()
                                .commentChildren())
                .map(commentChildren -> TextHelper.flatten(commentChildren, 0))
                .map(comments -> {
                    if (isSingleCommentThread) {
                        comments.set(0, comments.get(0).withKind(Comment.kind.SINGLECOMMENTTOP));
                    }
                    return comments;
                });
    }

    private Function<PostListing, Pair<String, List<Post>>> funcFormatPostData = postListing -> {
        String nextPageId = postListing.data().nextPage();

        return Observable.zip(Observable.fromArray(nextPageId),
                getAndFormatPosts(Observable.fromArray(postListing)),
                Pair::new)
                .blockingSingle();
    };

    private Observable<List<Post>> getAndFormatPosts(Observable<PostListing> postListingObservable) {
        return postListingObservable
                .map(postListing -> postListing.data().Posts())
                .map(Observable::fromIterable)
                .flatMap(TextHelper.funcFormatPostResponse)
                .toList().toObservable();
    }

    private Function<SubredditListing, Pair<String, List<SearchResult>>> formatSearchData = subredditListing -> {
        Observable<SubredditListing> subredditListingObservable = Observable.fromArray(subredditListing);
        return Observable.zip(Observable.fromArray(subredditListing.data().after()),
                getAndFormatSearchResults(subredditListingObservable),
                Pair::new)
                .blockingSingle();
    };

    private Observable<List<SearchResult>> getAndFormatSearchResults(Observable<SubredditListing> subredditListing) {
        return subredditListing.map(subredditListing1 -> Observable.fromIterable(subredditListing1.data().children()))
                .flatMap(TextHelper::formatSearchResult)
                .toList().toObservable();
    }
}

File: app/src/main/java/torille/fi/lurkforreddit/data/remote/RedditRemoteDataSource.kt
package torille.fi.lurkforreddit.data.remote

import android.text.TextUtils
import com.google.gson.stream.JsonReader
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import torille.fi.lurkforreddit.data.RedditDataSource
import torille.fi.lurkforreddit.data.RedditService
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentListing
import torille.fi.lurkforreddit.data.models.jsonResponses.PostListing
import torille.fi.lurkforreddit.data.models.jsonResponses.PostResponse
import torille.fi.lurkforreddit.data.models.jsonResponses.SubredditListing
import torille.fi.lurkforreddit.data.models.view.*
import torille.fi.lurkforreddit.di.scope.Remote
import torille.fi.lurkforreddit.utils.CommentsStreamingParser
import torille.fi.lurkforreddit.utils.Store
import torille.fi.lurkforreddit.utils.TextHelper
import java.io.InputStreamReader
import javax.inject.Inject

/**
 * Implementation of Reddit API
 */
@Remote
class RedditRemoteDataSource @Inject
internal constructor(
    private val redditApi: RedditService.Reddit,
    private val store: Store,
    private val commentsStreamingParser: CommentsStreamingParser
) : RedditDataSource {

    override fun saveSubreddits(subreddit: List<Subreddit>) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun saveSubreddit(subreddit: Subreddit) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getSubreddits(): Flowable<List<Subreddit>> {
        Timber.d("Fetching subs!")
        val subreddits: Flowable<SubredditListing>
        if (store.isLoggedIn) {
            Timber.d("Was logged in, getting personal subreddits")
            subreddits = redditApi.getMySubreddits(200)

            return Flowable.zip(fetchSubreddits(subreddits), getUserMultireddits(),
                BiFunction({ s: List<Subreddit>, m: List<Subreddit> ->
                    s.plus(m)
                })
            )
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
        } else {
            Timber.d("Was not logged in, getting default subreddits")
            subreddits = redditApi.getDefaultSubreddits(100)

            return fetchSubreddits(subreddits)
                .observeOn(AndroidSchedulers.mainThread())
        }
    }

    override fun getSubredditPosts(subredditUrl: String): Observable<kotlin.Pair<String, List<Post>>> {
        return redditApi
            .getSubreddit(subredditUrl)
            .observeOn(Schedulers.computation())
            .flatMap(funcFormatPostData)
    }

    override fun getMoreSubredditPosts(
        subredditUrl: String,
        nextpageId: String
    ): Observable<kotlin.Pair<String, List<Post>>> {
        return redditApi.getSubredditNextPage(subredditUrl, nextpageId)
            .observeOn(Schedulers.computation())
            .flatMap(funcFormatPostData)
    }

    override fun refreshData() {
        //empty
    }

    override fun getCommentsForPost(
        permaLinkUrl: String,
        isSingleCommentThread: Boolean
    ): Observable<PostAndComments> {
        return redditApi.getComments(permaLinkUrl)
            .map { responseBody ->
                responseBody.byteStream().use { stream ->
                    InputStreamReader(stream, "UTF-8").use { input ->
                        JsonReader(input).use { reader ->
                            commentsStreamingParser.readCommentListingArray(reader)
                        }
                    }
                }
            }
            .observeOn(Schedulers.computation())
            .concatMap { commentListings ->
                val commentListingsObservable = Observable.fromArray(commentListings)
                Observable.zip<Post, List<Comment>, PostAndComments>(
                    getPost(commentListingsObservable),
                    getFormattedComments(commentListingsObservable, isSingleCommentThread),
                    BiFunction { post, comments -> PostAndComments(post, comments) })
            }
    }

    override fun getMoreCommentsForPostAt(
        childCommentIds: List<String>,
        linkId: String,
        commentLevel: Int
    ): Observable<List<Comment>> {
        return redditApi.getMoreComments(
            linkId,
            TextUtils.join(
                ",",
                childCommentIds
            ),
            "json"
        )
            .observeOn(Schedulers.computation())
            .map { responseBody ->
                responseBody.byteStream().use { stream ->
                    InputStreamReader(stream, "UTF-8").use { input ->
                        JsonReader(input).use { reader ->
                            commentsStreamingParser.readMoreComments(reader)
                        }
                    }
                }
            }
            .map { commentChildren ->
                TextHelper.flattenAdditionalComments(
                    commentChildren,
                    commentLevel
                )
            }
    }

    override fun getSearchResults(query: String): Flowable<kotlin.Pair<String, List<SearchResult>>> {

        return redditApi.searchSubreddits(query, "relevance")
            .observeOn(Schedulers.computation())
            .concatMap(formatSearchData)
    }

    override fun getMoreSearchResults(
        query: String,
        after: String
    ): Flowable<kotlin.Pair<String, List<SearchResult>>> {
        return redditApi.searchSubredditsNextPage(query, "relevance", after)
            .observeOn(Schedulers.computation())
            .concatMap(formatSearchData)
    }

    private fun getUserMultireddits(): Flowable<List<Subreddit>> {
        return redditApi.getUserMultireddits()
            .observeOn(Schedulers.computation())
            .map { multiredditListingArray ->
                multiredditListingArray.map { (multireddit) ->
                    Subreddit(
                        subId = multireddit.name,
                        url = multireddit.pathUrl,
                        keyColor = multireddit.keyColor,
                        displayName = multireddit.displayName
                    )
                }
            }
    }


    private fun fetchSubreddits(listingFlowable: Flowable<SubredditListing>): Flowable<List<Subreddit>> {
        return listingFlowable
            .subscribeOn(Schedulers.computation())
            .map { subredditListing -> Flowable.fromIterable(subredditListing.data.children) }
            .flatMap { TextHelper.formatSubreddit(it) }
            .toSortedList { subreddit1, subreddit2 ->
                subreddit1.displayName.compareTo(
                    subreddit2.displayName,
                    ignoreCase = true
                )
            }
            .toFlowable()
    }

    private fun getPost(observable: Observable<List<CommentListing>>): Observable<Post> {
        return observable
            .map { commentListings ->
                commentListings[0]
                    .commentData
                    .commentChildren[0]
                    .originalPost!!
            }
            .map(TextHelper.funcFormatPost)
    }

    private fun getFormattedComments(
        observable: Observable<List<CommentListing>>,
        isSingleCommentThread: Boolean
    ): Observable<List<Comment>> {

        return observable
            .subscribeOn(Schedulers.computation())
            .map { commentListings ->
                commentListings[1]
                    .commentData
                    .commentChildren
            }
            .map { TextHelper.flatten(it, 0) }
            .map { comments ->
                if (isSingleCommentThread) {
                    val commentsMutable = comments.toMutableList()
                    commentsMutable[0] = commentsMutable[0].copy(kind = kind.SINGLECOMMENTTOP)
                    commentsMutable.toList()
                } else {
                    comments
                }
            }
    }

    private val funcFormatPostData = { postListing: PostListing ->

        Observable.zip(Observable.fromArray<String>(postListing.data.nextPage),
            getAndFormatPosts(Observable.fromArray(postListing)),
            BiFunction<String, List<Post>, kotlin.Pair<String, List<Post>>> { first, second ->
                kotlin.Pair(first, second)
            })
    }

    private fun getAndFormatPosts(postListingObservable: Observable<PostListing>): Observable<List<Post>> {
        return postListingObservable
            .map { postListing -> postListing.data.Posts }
            .map { Observable.fromIterable(it) }
            .flatMap({ postResponseObservable: Observable<PostResponse> ->
                postResponseObservable
                    .map({ it.postDetails })
                    .map(TextHelper.funcFormatPost)
            })
            .toList().toObservable()
    }

    private val formatSearchData = { subredditListing: SubredditListing ->
        val subredditListingFlowable = Flowable.fromArray<SubredditListing>(subredditListing)
        Flowable.zip(Flowable.fromArray(subredditListing.data.after),
            getAndFormatSearchResults(subredditListingFlowable),
            BiFunction<String, List<SearchResult>, kotlin.Pair<String, List<SearchResult>>> { first, second ->
                kotlin.Pair(first, second)
            })
    }

    private fun getAndFormatSearchResults(subredditListing: Flowable<SubredditListing>): Flowable<List<SearchResult>> {
        return subredditListing.map { subredditListing1 -> Flowable.fromIterable(subredditListing1.data.children) }
            .flatMap { TextHelper.formatSearchResult(it) }
            .toList().toFlowable()
    }
}
